
# Retail Store Sales — EDA & Cleaning Notebook

This notebook consolidates your three Python scripts into a single, streamlined workflow for:
1) **Extraction / Loading & Quick Audit**
2) **Exploratory Data Analysis (EDA)**
3) **Cleaning + Intelligent Filling + Exports**

**Key improvements vs. the original scripts:**
- Single CSV read (no repeated I/O).
- Robust type coercions and validation, without overwriting the original raw file.
- All plots rendered with **matplotlib only** (no seaborn), which is simpler and self-contained.
- Reusable helper functions to avoid duplication.
- Clear save locations for artifacts (JSON summaries, cleaned CSVs, and charts).


In [1]:

# Imports & settings
import os
import json
import warnings
from datetime import datetime

import seaborn as sns
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick
from matplotlib.ticker import FuncFormatter
warnings.filterwarnings('ignore')

# Pandas display prefs (safe for notebooks)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

# Paths
PROJECT_DIR = r"C:\Users\matth\etl_online_retail_project"
print(PROJECT_DIR)
DATA_DIR = os.path.join(PROJECT_DIR, 'data')
RAW_DATA_PATH = os.path.join(DATA_DIR, 'retail_store_sales.csv')

# Output directories
ARTIFACTS_DIR = os.path.join(DATA_DIR, 'artifacts')
os.makedirs(ARTIFACTS_DIR, exist_ok=True)

# Common column groups
NUMERIC_COLS = ['Price Per Unit', 'Quantity', 'Total Spent']
CATEGORICAL_COLS = ['Category', 'Payment Method', 'Location', 'Discount Applied']
DATE_COL = 'Transaction Date'
KEY_COL = 'Transaction ID'  # used for duplicate checks, if present


C:\Users\matth\etl_online_retail_project


In [2]:

# -----------------------------
# Helper functions
# -----------------------------

def safe_read_csv(path):
    """Read CSV with UTF-8 fallback to latin-1, without overwriting the raw file."""
    try:
        return pd.read_csv(path)
    except UnicodeDecodeError:
        return pd.read_csv(path, encoding='latin1')

def coerce_types(df):
    """Coerce expected dtypes for EDA/cleaning."""
    df = df.copy()
    # Date
    if DATE_COL in df.columns:
        df[DATE_COL] = pd.to_datetime(df[DATE_COL], errors='coerce')
    # Numerics
    for col in NUMERIC_COLS:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors='coerce')
    # Discount Applied to boolean (with NaN preserved)
    if 'Discount Applied' in df.columns:
        # Map strings to booleans where clear, preserve NaN
        ser = df['Discount Applied'].astype(str).str.strip().str.lower()
        mapped = ser.map({'true': True, 'false': False})
        df['Discount Applied'] = mapped.where(~ser.isin(['nan', 'none', '']), np.nan)
    return df

def quick_audit(df):
    """Return a concise dict of dataset stats."""
    mem_mb = df.memory_usage(deep=True).sum() / 1024**2
    missing_values = df.isnull().sum()
    missing_pct = (missing_values / len(df)) * 100
    dtypes_count = {str(dt): int((df.dtypes == dt).sum()) for dt in df.dtypes.unique()}
    return {
        'shape': df.shape,
        'memory_mb': round(mem_mb, 1),
        'dtype_counts': dtypes_count,
        'total_missing_cells': int(missing_values.sum()),
        'missing_pct': float((missing_values.sum() / (len(df) * len(df.columns))) * 100),
        'missing_df': pd.DataFrame({
            'Column': missing_values.index,
            'Missing_Count': missing_values.values,
            'Missing_Percentage': missing_pct.values
        }).sort_values('Missing_Count', ascending=False)
    }

def validate_financials(df, tolerance=0.01):
    """Check Price*Quantity vs Total Spent discrepancies."""
    if not all(col in df.columns for col in NUMERIC_COLS):
        return 0
    complete = df.dropna(subset=NUMERIC_COLS)
    if complete.empty:
        return 0
    calc_total = complete['Price Per Unit'] * complete['Quantity']
    discrep = (calc_total - complete['Total Spent']).abs() > tolerance
    return int(discrep.sum())

def detect_date_anomalies(df):
    """Future/past date checks based on current year and >= 2020 lower bound."""
    if DATE_COL not in df.columns:
        return 0, 0
    current_year = datetime.now().year
    future = df[df[DATE_COL].dt.year > current_year]
    past = df[df[DATE_COL].dt.year < 2020]
    return len(future), len(past)

def save_json(obj, filename):
    path = os.path.join(ARTIFACTS_DIR, filename)
    with open(path, 'w') as f:
        json.dump(obj, f, indent=2, default=str)
    return path

def plot_missing(missing_df, filename='missing_values.png'):
    miss = missing_df[missing_df['Missing_Count'] > 0]
    if miss.empty:
        return None
    plt.figure(figsize=(10, 6))
    plt.bar(miss['Column'], miss['Missing_Percentage'])
    plt.title('Missing Values by Column')
    plt.xlabel('Column')
    plt.ylabel('Missing Percentage')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    out = os.path.join(ARTIFACTS_DIR, filename)
    plt.savefig(out, dpi=300, bbox_inches='tight')
   # plt.show()   
    return out

def plot_outliers(df, filename='outlier_detection.png'):
    cols = [c for c in NUMERIC_COLS if c in df.columns]
    if not cols:
        return None
    n = len(cols)
    plt.figure(figsize=(5*n, 5))
    for i, col in enumerate(cols, 1):
        plt.subplot(1, n, i)
        df.boxplot(column=col)
        plt.title(f'{col} Distribution')
    plt.tight_layout()
    out = os.path.join(ARTIFACTS_DIR, filename)
    plt.savefig(out, dpi=300, bbox_inches='tight')
  #  plt.show()   
    return out

def plot_categorical_distributions(df, filename='categorical_distributions.png'):
    cols = [c for c in CATEGORICAL_COLS if c in df.columns]
    if not cols:
        return None
    plt.figure(figsize=(15, 10))
    for i, col in enumerate(cols[:4], 1):
        plt.subplot(2, 2, i)
        vc = df[col].value_counts().head(10)
        plt.bar(vc.index.astype(str), vc.values)
        plt.title(f'{col} Distribution')
        plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    out = os.path.join(ARTIFACTS_DIR, filename)
    plt.savefig(out, dpi=300, bbox_inches='tight')
   # plt.show()   
    return out
def to_cents(x):
    """Convert a numeric dollar value to integer cents, handling NaN safely."""
    if pd.isna(x): 
        return 0
    return int(round(float(x) * 100.0))

def norm_channel(loc_value: str) -> str:
    """Normalize location/channel into 'Online' or 'In-Store'."""
    if pd.isna(loc_value):
        return "Online"
    s = str(loc_value).strip().lower()
    return "In-Store" if s in ("in-store","instore","in store","in-store") else "Online"

def boolish_to_int(x):
    """Map True/False/yes/no/1/0 strings to 1/0 for SQLite."""
    if isinstance(x, (bool, np.bool_)):
        return int(x)
    if pd.isna(x):
        return 0
    s = str(x).strip().lower()
    return 1 if s in ("true","1","yes","y","t") else 0

## 1) Ingest Raw Retail Data

In [3]:

df_raw = safe_read_csv(RAW_DATA_PATH)
print(f"Data shape: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns")

# Show columns at a glance
df_raw.head()


Data shape: 12,575 rows × 11 columns


Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  
0             True  
1             True  
2            False  
3              NaN  
4            False

## 2) Data Quality & Validation

In [4]:

df = coerce_types(df_raw)
audit = quick_audit(df)

print(f"Memory usage: {audit['memory_mb']} MB\nDtype counts: {audit['dtype_counts']}\n"
      f"Total missing cells: {audit['total_missing_cells']:,} "
      f"({audit['missing_pct']:.1f}%)") 

print(audit['missing_df'].head(20))

future_cnt, past_cnt = detect_date_anomalies(df)
discrep_cnt = validate_financials(df)

exact_duplicates = int(df.duplicated().sum())
key_duplicates = int(df.duplicated(subset=[KEY_COL]).sum()) if KEY_COL in df.columns else 0

quality_flags = {
    'future_dates': future_cnt,
    'pre_2020_dates': past_cnt,
    'financial_discrepancies': discrep_cnt,
    'exact_duplicates': exact_duplicates,
    'key_duplicates': key_duplicates
}

print(quality_flags)

Memory usage: 5.6 MB
Dtype counts: {'object': 7, 'float64': 3, 'datetime64[ns]': 1}
Total missing cells: 7,229 (5.2%)
              Column  Missing_Count  Missing_Percentage
10  Discount Applied           4199           33.391650
3               Item           1213            9.646123
4     Price Per Unit            609            4.842942
5           Quantity            604            4.803181
6        Total Spent            604            4.803181
0     Transaction ID              0            0.000000
1        Customer ID              0            0.000000
2           Category              0            0.000000
7     Payment Method              0            0.000000
8           Location              0            0.000000
9   Transaction Date              0            0.000000
{'future_dates': 0, 'pre_2020_dates': 0, 'financial_discrepancies': 0, 'exact_duplicates': 0, 'key_duplicates': 0}


## 3) Automated Cleaning Pipeline

In [5]:
# === Cleaning aligned to the outlined practices (uses provided helpers) ===
TOL = 0.01  # 1-cent tolerance for financial checks

# Start from coerced types to ensure safe operations
df_clean = coerce_types(df)
initial_records = len(df_clean)

# ---------------------------
# 1) DATES
# ---------------------------
# Quarantine invalid/null dates; keep for QA review
quarantine_dates = pd.DataFrame(columns=df_clean.columns)
if DATE_COL in df_clean.columns:
    invalid_date_mask = df_clean[DATE_COL].isna()
    if invalid_date_mask.any():
        quarantine_dates = pd.concat([quarantine_dates, df_clean[invalid_date_mask]], ignore_index=True)
        df_clean = df_clean[~invalid_date_mask].copy()

    # Flag future & pre-2020 for business validation
    current_year = datetime.now().year
    df_clean['flag_future_date'] = df_clean[DATE_COL].dt.year > current_year
    df_clean['flag_pre2020_date'] = df_clean[DATE_COL].dt.year < 2020

    # Temporal columns for downstream transformations
    df_clean['Year'] = df_clean[DATE_COL].dt.year
    df_clean['Month'] = df_clean[DATE_COL].dt.month
    df_clean['Day'] = df_clean[DATE_COL].dt.day
    df_clean['DayOfWeek'] = df_clean[DATE_COL].dt.day_name()
else:
    df_clean['flag_future_date'] = False
    df_clean['flag_pre2020_date'] = False

# ---------------------------
# 2) NUMERICS
# ---------------------------
# Negative flags
for nc in [c for c in NUMERIC_COLS if c in df_clean.columns]:
    df_clean[f'flag_negative_{nc}'] = df_clean[nc] < 0

# Internal consistency: Total Spent ≈ Price × Quantity (only where all 3 non-null)
if all(c in df_clean.columns for c in NUMERIC_COLS):
    complete_num = df_clean[NUMERIC_COLS].notna().all(axis=1)
    expected_total = df_clean.loc[complete_num, 'Price Per Unit'] * df_clean.loc[complete_num, 'Quantity']
    discrep_mask = pd.Series(False, index=df_clean.index)
    discrep_mask.loc[complete_num] = (expected_total - df_clean.loc[complete_num, 'Total Spent']).abs() > TOL
    df_clean['flag_discrepancy_total'] = discrep_mask
else:
    df_clean['flag_discrepancy_total'] = False
# --- PRE-REPAIR DIAGNOSTICS ---                                         # <<< NEW
pre_dm = df_clean.get('flag_discount_mismatch', pd.Series(False, index=df_clean.index)).fillna(False)   # <<< NEW
pre_dm_count = int(pre_dm.sum())                                                                            # <<< NEW
pre_flag_cols = [c for c in df_clean.columns if c.startswith('flag_')]                                      # <<< NEW
pre_flag_counts = df_clean[pre_flag_cols].sum().sort_values(ascending=False) if pre_flag_cols else None     # <<< NEW
# Intelligent backfill when exactly one of the three is missing
int_fill_actions = {}
if all(c in df_clean.columns for c in NUMERIC_COLS):
    # Fill Total Spent if Price & Quantity present
    m_total = df_clean['Total Spent'].isna() & df_clean['Price Per Unit'].notna() & df_clean['Quantity'].notna()
    df_clean.loc[m_total, 'Total Spent'] = df_clean.loc[m_total, 'Price Per Unit'] * df_clean.loc[m_total, 'Quantity']
    int_fill_actions['totals_from_price_qty'] = int(m_total.sum())

    # Fill Price Per Unit if Total & Quantity present (quantity != 0)
    m_price = df_clean['Price Per Unit'].isna() & df_clean['Total Spent'].notna() & df_clean['Quantity'].notna() & (df_clean['Quantity'] != 0)
    df_clean.loc[m_price, 'Price Per Unit'] = df_clean.loc[m_price, 'Total Spent'] / df_clean.loc[m_price, 'Quantity']
    int_fill_actions['price_from_total_qty'] = int(m_price.sum())

    # Fill Quantity if Total & Price present (price != 0)
    m_qty = df_clean['Quantity'].isna() & df_clean['Total Spent'].notna() & df_clean['Price Per Unit'].notna() & (df_clean['Price Per Unit'] != 0)
    df_clean.loc[m_qty, 'Quantity'] = df_clean.loc[m_qty, 'Total Spent'] / df_clean.loc[m_qty, 'Price Per Unit']
    int_fill_actions['qty_from_total_price'] = int(m_qty.sum())

# ---------------------------
# 3) DISCOUNTS
# ---------------------------
# coerce_types already normalized 'Discount Applied' -> boolean/NaN
# Ensure metric columns exist
if 'Discount_Amount' not in df_clean.columns:
    df_clean['Discount_Amount'] = 0.0
if 'Discount_Percentage' not in df_clean.columns:
    df_clean['Discount_Percentage'] = 0.0

filling_actions = {}
if all(c in df_clean.columns for c in NUMERIC_COLS):
    if 'Discount Applied' in df_clean.columns:
        # Infer discount where flag is NaN but math implies discount
        missing_discounts = int(df_clean['Discount Applied'].isna().sum())
        mask = df_clean['Discount Applied'].isna() & df_clean[NUMERIC_COLS].notna().all(axis=1)
        exp_total = df_clean.loc[mask, 'Price Per Unit'] * df_clean.loc[mask, 'Quantity']
        diff = (exp_total - df_clean.loc[mask, 'Total Spent'])
        disc_mask = diff > TOL
        no_disc_mask = ~disc_mask

        df_clean.loc[mask & disc_mask, 'Discount Applied'] = True
        df_clean.loc[mask & no_disc_mask, 'Discount Applied'] = False

        df_clean.loc[mask & disc_mask, 'Discount_Amount'] = diff[disc_mask].values
        with np.errstate(divide='ignore', invalid='ignore'):
            df_clean.loc[mask & disc_mask, 'Discount_Percentage'] = np.where(
                exp_total[disc_mask].values != 0,
                (df_clean.loc[mask & disc_mask, 'Discount_Amount'] / exp_total[disc_mask].values) * 100,
                0.0
            )
        filling_actions['discounts_inferred'] = int((mask & disc_mask).sum())
    else:
        missing_discounts = 0

    # Compute for rows with Discount Applied == True (idempotent)
    if 'Discount Applied' in df_clean.columns:
        true_mask = df_clean['Discount Applied'] == True
        exp_total_true = df_clean.loc[true_mask, 'Price Per Unit'] * df_clean.loc[true_mask, 'Quantity']
        diff_true = (exp_total_true - df_clean.loc[true_mask, 'Total Spent']).clip(lower=0)
        df_clean.loc[true_mask, 'Discount_Amount'] = diff_true.values
        with np.errstate(divide='ignore', invalid='ignore'):
            df_clean.loc[true_mask, 'Discount_Percentage'] = np.where(
                exp_total_true.values != 0,
                (df_clean.loc[true_mask, 'Discount_Amount'] / exp_total_true.values) * 100,
                0.0
            )

        # Flag mismatches: boolean vs. math
        df_clean['flag_discount_mismatch'] = False
        mismatch_true = true_mask & (df_clean['Discount_Amount'] <= TOL)  # True but no discount by math
        implied_disc_mask = (df_clean['Discount Applied'] == False) & df_clean[NUMERIC_COLS].notna().all(axis=1)
        implied_exp = df_clean.loc[implied_disc_mask, 'Price Per Unit'] * df_clean.loc[implied_disc_mask, 'Quantity']
        implied_disc_mask = implied_disc_mask & ((implied_exp - df_clean.loc[implied_disc_mask, 'Total Spent']) > TOL)
        df_clean.loc[mismatch_true | implied_disc_mask, 'flag_discount_mismatch'] = True
else:
    df_clean['flag_discount_mismatch'] = False
# --- Capture PRE-REPAIR diagnostics (BEFORE the repair block) ---
# Basic overlap counts
pre_dm = df_clean.get('flag_discount_mismatch', pd.Series(False, index=df_clean.index)).fillna(False)
pre_disc = df_clean.get('flag_discrepancy_total', pd.Series(False, index=df_clean.index)).fillna(False)

pre_dm_count = int(pre_dm.sum())
pre_disc_count = int(pre_disc.sum())
pre_overlap_both = int((pre_dm & pre_disc).sum())
pre_only_dm = pre_dm_count - pre_overlap_both
pre_only_disc = pre_disc_count - pre_overlap_both

# Per-flag totals (any column that starts with 'flag_')
pre_flag_cols = [c for c in df_clean.columns if c.startswith('flag_')]
pre_flag_counts = (
    df_clean[pre_flag_cols].sum().sort_values(ascending=False)
    if pre_flag_cols else None
)
# ---------------------------
# 3b) REPAIR DISCOUNT MISMATCH ROWS (make analysis-ready)
# ---------------------------
if 'flag_discount_mismatch' in df_clean.columns and all(c in df_clean.columns for c in NUMERIC_COLS):
    dm = df_clean['flag_discount_mismatch'].fillna(False) & df_clean[NUMERIC_COLS].notna().all(axis=1)
    fix_count = int(dm.sum())
    if fix_count:
        exp = df_clean.loc[dm, 'Price Per Unit'] * df_clean.loc[dm, 'Quantity']
        diff = (exp - df_clean.loc[dm, 'Total Spent'])

        as_disc = diff > TOL       # implied discount
        as_nodisc = ~as_disc       # no discount implied

        # Apply repaired values
        df_clean.loc[dm & as_disc, 'Discount Applied'] = True
        df_clean.loc[dm & as_disc, 'Discount_Amount'] = diff[as_disc].values
        with np.errstate(divide='ignore', invalid='ignore'):
            df_clean.loc[dm & as_disc, 'Discount_Percentage'] = np.where(
                exp[as_disc].values != 0,
                (df_clean.loc[dm & as_disc, 'Discount_Amount'] / exp[as_disc].values) * 100,
                0.0
            )

        df_clean.loc[dm & as_nodisc, 'Discount Applied'] = False
        df_clean.loc[dm & as_nodisc, ['Discount_Amount', 'Discount_Percentage']] = 0.0

        # mismatch resolved -> clear the flag
        df_clean.loc[dm, 'flag_discount_mismatch'] = False

        print(f"Repaired {fix_count:,} rows with discount mismatch by inferring discount from Price×Quantity vs Total.")

# ---------------------------
# 4) CATEGORICALS (standardize)
# ---------------------------
def _norm_str(s):
    return str(s).strip().lower() if pd.notna(s) else s

# Minimal controlled vocab examples — expand per your domain
category_map = {
    'electronics': 'Electronics',
    'grocery': 'Grocery',
    'clothing': 'Clothing',
}
payment_map = {
    'credit': 'Credit Card',
    'credit card': 'Credit Card',
    'debit': 'Debit Card',
    'debit card': 'Debit Card',
    'cash': 'Cash',
    'digital wallet': 'Digital Wallet',
}
location_map = {
    'nyc': 'New York',
    'new york': 'New York',
    'sf': 'San Francisco',
    'san francisco': 'San Francisco',
}

if 'Category' in df_clean.columns:
    df_clean['Category_std'] = df_clean['Category'].map(lambda x: category_map.get(_norm_str(x), str(x).strip().title()))
if 'Payment Method' in df_clean.columns:
    df_clean['Payment Method_std'] = df_clean['Payment Method'].map(lambda x: payment_map.get(_norm_str(x), str(x).strip().title()))
if 'Location' in df_clean.columns:
    df_clean['Location_std'] = df_clean['Location'].map(lambda x: location_map.get(_norm_str(x), str(x).strip().title()))

# ---------------------------
# 5) DUPLICATES
# ---------------------------
removed_exact_dups = int(df_clean.duplicated().sum())
df_clean = df_clean.drop_duplicates()

if KEY_COL in df_clean.columns:
    df_clean['flag_txn_duplicate'] = df_clean.duplicated(subset=[KEY_COL], keep='first')
else:
    df_clean['flag_txn_duplicate'] = False

# ---------------------------
# Final splits (analysis-ready vs QA)
# ---------------------------
# Build flag lists AFTER all flags/repairs
hard_flags = [
    'flag_txn_duplicate',        # corrupt key
    'flag_discrepancy_total',    # core arithmetic inconsistency
] + [c for c in df_clean.columns if c.startswith('flag_negative_')]  # adjust per returns policy

soft_flags = [
    'flag_discount_mismatch',    # mostly cleared by repair step
    'flag_future_date',
    'flag_pre2020_date',
]

# Masks
no_hard_flags_mask = (~pd.concat([df_clean[c] for c in hard_flags], axis=1).any(axis=1)) if hard_flags else pd.Series(True, index=df_clean.index)
critical_columns = ['Item', 'Price Per Unit', 'Quantity', 'Total Spent']
present_critical = [c for c in critical_columns if c in df_clean.columns]
complete_mask = df_clean[present_critical].notna().all(axis=1) if present_critical else pd.Series(False, index=df_clean.index)

analysis_ready_mask = complete_mask & no_hard_flags_mask
df_final_complete = df_clean[analysis_ready_mask].copy()
df_final_incomplete = df_clean[~analysis_ready_mask].copy()

# --- Drop flag columns for analysis-ready & full-cleaned outputs ---
flag_cols = [c for c in df_clean.columns if c.startswith('flag_')]
df_final_complete_noflags = df_final_complete.drop(columns=flag_cols, errors='ignore')
df_clean_noflags = df_clean.drop(columns=flag_cols, errors='ignore')

# --- POST-repair diagnostics ---
post_dm = df_clean.get('flag_discount_mismatch', pd.Series(False, index=df_clean.index)).fillna(False)
post_disc = df_clean.get('flag_discrepancy_total', pd.Series(False, index=df_clean.index)).fillna(False)

post_dm_count = int(post_dm.sum())
post_disc_count = int(post_disc.sum())
post_overlap_both = int((post_dm & post_disc).sum())
post_only_dm = post_dm_count - post_overlap_both
post_only_disc = post_disc_count - post_overlap_both

post_flag_cols = [c for c in df_clean.columns if c.startswith('flag_')]
post_flag_counts = (
    df_clean[post_flag_cols].sum().sort_values(ascending=False)
    if post_flag_cols else None
)

# ---------------------------
# End Summary
# ---------------------------
print("\n=== CLEANING SUMMARY ===")
print(f"Original rows:                    {initial_records:,}")
print(f"After exact-dup drop:             {len(df_clean):,} (removed {removed_exact_dups:,})")
print(f"Analysis-ready rows (post-repair): {len(df_final_complete):,}")
print(f"Incomplete/QA rows (post-repair):  {len(df_final_incomplete):,}")
print("Intelligent fill:", int_fill_actions)


# Flag counts (pre/post)
if pre_flag_counts is not None:
    print("\n[Pre-repair] Rows flagged by each rule:")
    print(pre_flag_counts.rename(lambda c: c.replace('flag_', '')))
if post_flag_counts is not None:
    print("\n[Post-repair] Rows flagged by each rule:")
    print(post_flag_counts.rename(lambda c: c.replace('flag_', '')))

# Improvements summary
resolved_mismatches = pre_dm_count - post_dm_count
resolved_pct = (resolved_mismatches / pre_dm_count * 100) if pre_dm_count else 0.0
print("\n=== Improvements ===")
print(f"Discount mismatch rows repaired: {resolved_mismatches:,} of {pre_dm_count:,} "
      f"({resolved_pct:.1f}%) — via inferred discount from Price×Quantity vs Total.")
print(f"Rows directly updated in repair step: {fix_count:,} (fields normalized: "
      f"'Discount Applied', 'Discount_Amount', 'Discount_Percentage').")

# Report flag removal for outputs
print(f"\nFlag columns removed from analysis-ready/full-cleaned outputs: {flag_cols}")



Repaired 4,019 rows with discount mismatch by inferring discount from Price×Quantity vs Total.

=== CLEANING SUMMARY ===
Original rows:                    12,575
After exact-dup drop:             12,575 (removed 0)
Analysis-ready rows (post-repair): 11,362
Incomplete/QA rows (post-repair):  1,213
Intelligent fill: {'totals_from_price_qty': 0, 'price_from_total_qty': 609, 'qty_from_total_price': 0}

[Pre-repair] Rows flagged by each rule:
discount_mismatch          4019
future_date                   0
pre2020_date                  0
negative_Price Per Unit       0
negative_Quantity             0
negative_Total Spent          0
discrepancy_total             0
dtype: int64

[Post-repair] Rows flagged by each rule:
future_date                0
pre2020_date               0
negative_Price Per Unit    0
negative_Quantity          0
negative_Total Spent       0
discrepancy_total          0
discount_mismatch          0
txn_duplicate              0
dtype: int64

=== Improvements ===
Discount mis

## Reflections on Data Quality Checks: “Strict by Design”

Our cleaning pass was intentionally **watchful**: we tested for date anomalies, negative numerics, arithmetic discrepancies, discount mismatches, duplicates, and categorical inconsistencies. The post-run metrics show many flags at **zero** and 4,019 **discount mismatches** fully repaired by inference (Price×Quantity vs Total). That’s not wasted work—it’s **assurance**.

### Why “over-checking” is valuable even when most flags are zero
- **Proves absence, not assumption**: Zero counts mean we *verified* those risks aren’t present in this batch. That’s stronger than assuming they don’t exist.
- **Guardrails for future drift**: Data pipelines face schema changes, vendor updates, and human entry error. These checks act as tripwires—quiet today, but ready to catch tomorrow’s anomalies.
- **Auditability & reproducibility**: Clear flags and a repair log create a defensible record for stakeholders (finance, ops, compliance) of what we validated and when.
- **Targeted fixes where it matters**: The discount logic identified 4,019 rows and **repaired them deterministically**, improving analysis readiness without blanket imputation.
- **Business trust**: Downstream analyses (margins, promos, mix) depend on consistent price/quantity/total relationships. Validating and aligning these builds confidence in every KPI built on top.

### What the results imply about this dataset
- **Discounts were the primary friction point**, and our inference strategy resolved them completely, aligning math with semantics (`Discount Applied`, `Discount_Amount`, `%`).
- **Other risk vectors are clean (today)**: No systemic negatives, date anomalies, or arithmetic discrepancies. That’s a strong indicator of source stability.
- **Incomplete rows remain for QA not because we’re punitive**, but because we separate “analysis-ready” from “needs-clarification,” preventing silent data leakage into metrics.

### How this helps downstream (T & L)
- **Transformation-ready inputs**: With repaired discounts and clean numerics, aggregated facts (e.g., revenue by item/location/date) won’t be biased by hidden inconsistencies.
- **Reliable dimensions**: Standardized categoricals (`*_std`) make joins stable and dashboard slice-and-dice consistent.
- **Sustainable operations**: The same checks can run on new drops; zero today becomes signal tomorrow if something breaks upstream.

> **Bottom line:** Even when most checks “do nothing,” they deliver *proof of integrity*. The one area that needed attention—discounts—was fixed in a transparent, reproducible way, lifting **11,362** rows into analysis-ready with confidence and leaving a clear QA trail for the rest.
>
>  **Editors Note:** Additionally, this data set is from a simulated kaggle download so even though it is "messy" it is all messy in the exact same way unlike a real dataset. All rows that didnt have price data, also failed to have an item SKU. This caused our 609 rows where we repaired price data to ultimately still be excluded.


## 4) Save and Inspect cleaned data

In [6]:

complete_data_path = os.path.join(DATA_DIR, 'cleanedCompleteForDataAnalysis.csv')
incomplete_data_path = os.path.join(DATA_DIR, 'cleanedForDataQualityAnalysis.csv')
full_cleaned_path = os.path.join(DATA_DIR, 'cleaned_retail_store_sales.csv')

df_final_complete.to_csv(complete_data_path, index=False)
df_final_incomplete.to_csv(incomplete_data_path, index=False)
df_clean.to_csv(full_cleaned_path, index=False)

cleaning_summary = {
    'cleaning_date': datetime.now().isoformat(),
    'records': {
        'original': initial_records,
        'after_exact_dup_drop': len(df_clean),
        'removed_exact_dups': removed_exact_dups,
        'analysis_ready': len(df_final_complete),
        'incomplete_or_QA': len(df_final_incomplete)
    },
    'cleaning_actions': {
        'invalid_dates_removed': int(invalid_dates) if 'invalid_dates' in locals() else 0,     # <<< CHANGED (guard)
        'missing_discounts_seen': int(missing_discounts) if 'missing_discounts' in locals() else 0,
        'intelligent_fill': int_fill_actions,
        'rows_repaired_in_discount_fix': fix_count,                                            # <<< NEW
        'fields_normalized': ['Discount Applied', 'Discount_Amount', 'Discount_Percentage']    # <<< NEW
    },
    'discount_diagnostics': {                                                                  # <<< NEW
        'pre_repair': {
            'mismatch_total': pre_dm_count,
            'flag_counts': pre_flag_counts.to_dict() if pre_flag_counts is not None else {}
        },
        'post_repair': {
            'mismatch_total': post_dm_count,
            'flag_counts': post_flag_counts.to_dict() if post_flag_counts is not None else {}
        },
        'improvements': {
            'resolved_mismatches': (pre_dm_count - post_dm_count),
            'percent_resolved': (100.0 * (pre_dm_count - post_dm_count) / pre_dm_count) if pre_dm_count else 0.0
        }
    },
    'notes': {
        'flags_dropped_in_outputs': [c for c in flag_cols],                                    # <<< NEW
        'outputs': {
            'complete_no_flags': os.path.basename(complete_data_path),
            'incomplete_with_flags': os.path.basename(incomplete_data_path),
            'full_cleaned_no_flags': os.path.basename(full_cleaned_path)
        }
    }
}
clean_json_path = save_json(cleaning_summary, 'cleaning_summary.json')
print(f" - Cleaning summary JSON:          {clean_json_path}")                                 # <<< NEW


clean_json_path = save_json(cleaning_summary, 'cleaning_summary.json')

print('Saved:')
#print(' -', complete_data_path)
#print(' -', incomplete_data_path)
#print(' -', full_cleaned_path)
#print(' -', clean_json_path)

df_clean_noflags.head(10)

 - Cleaning summary JSON:          C:\Users\matth\etl_online_retail_project\data\artifacts\cleaning_summary.json
Saved:


Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
5    TXN_7482416     CUST_09     Patisserie           NaN            20.0   
6    TXN_3652209     CUST_07           Food   Item_1_FOOD             5.0   
7    TXN_1372952     CUST_21      Furniture           NaN            33.5   
8    TXN_9728486     CUST_23      Furniture   Item_16_FUR            27.5   
9    TXN_2722661     CUST_25       Butchers   Item_22_BUT            36.5   

   Quantity  Total Spent  Payment Method  Location Transaction Date  \
0      10.0        185.0  Digital Wallet    Online       2024-04-08   
1       9.0        261.0  Digital Wallet    Online       2023-07-23   
2       2.0         43.0     Credit Card    Online       2022-10-05   
3       9.0        247.5     Credit Card    Online       2022-05-07   
4       7.0         87.5  Digital Wallet    Online       2022-10-02   
5      10.0        200.0     Credit Card    Online       2023-11-30   
6       8.0         40.0     Credit Card  In-store       2023-06-10   
7       NaN          NaN  Digital Wallet  In-store       2024-04-02   
8       1.0         27.5     Credit Card  In-store       2023-04-26   
9       3.0        109.5            Cash    Online       2024-03-14   

  Discount Applied  Year  Month  Day  DayOfWeek  Discount_Amount  \
0            False  2024      4    8     Monday              0.0   
1            False  2023      7   23     Sunday              0.0   
2            False  2022     10    5  Wednesday              0.0   
3            False  2022      5    7   Saturday              0.0   
4            False  2022     10    2     Sunday              0.0   
5            False  2023     11   30   Thursday              0.0   
6            False  2023      6   10   Saturday              0.0   
7             True  2024      4    2    Tuesday              NaN   
8            False  2023      4   26  Wednesday              0.0   
9            False  2024      3   14   Thursday              0.0   

   Discount_Percentage   Category_std Payment Method_std Location_std  
0                  0.0     Patisserie     Digital Wallet       Online  
1                  0.0  Milk Products     Digital Wallet       Online  
2                  0.0       Butchers        Credit Card       Online  
3                  0.0      Beverages        Credit Card       Online  
4                  0.0           Food     Digital Wallet       Online  
5                  0.0     Patisserie        Credit Card       Online  
6                  0.0           Food        Credit Card     In-Store  
7                  NaN      Furniture     Digital Wallet     In-Store  
8                  0.0      Furniture        Credit Card     In-Store  
9                  0.0       Butchers               Cash       Online

## 5) SQL Schema Design & ETL to SQLite

In [7]:
import sqlite3
DB_PATH = "sales.db"
df = df_clean_noflags

# Prefer standardized columns if present; otherwise use originals.
cat_col   = 'Category_std' if 'Category_std' in df.columns else 'Category'
pm_col    = 'Payment Method_std' if 'Payment Method_std' in df.columns else 'Payment Method'
loc_col   = 'Location_std' if 'Location_std' in df.columns else 'Location'
item_col  = 'Item'  # item_std not provided in sample; keep as-is

# Make a shallow copy with uniform column names for staging
stg = df.rename(columns={
    'Transaction ID':'transaction_id',
    'Customer ID':'customer_id',
    cat_col:'category',
    item_col:'item',
    'Price Per Unit':'price_per_unit',
    'Quantity':'quantity',
    'Total Spent':'total_spent',
    pm_col:'payment_method',
    loc_col:'location',
    'Transaction Date':'transaction_date',
    'Discount Applied':'discount_applied',
    'Discount_Amount':'discount_amount',
    'Discount_Percentage':'discount_percentage'
}).copy()

# Ensure expected columns exist (create if not)
for col, default in [
    ('discount_amount', 0.0),
    ('discount_percentage', 0.0),
]:
    if col not in stg.columns:
        stg[col] = default

# Normalize channel and booleans on the pandas side for consistency
stg['location'] = stg['location'].apply(norm_channel)
stg['discount_applied'] = stg['discount_applied'].apply(boolish_to_int)

# Parse dates to datetime (errors='coerce' puts bad dates as NaT, which we can quarantine if needed)
stg['transaction_date'] = pd.to_datetime(stg['transaction_date'], errors='coerce').dt.date

# Create a header staging table by deduping on transaction_id
# Keep the *last* non-null row per transaction (sortable by date then index)
stg_sorted = stg.sort_values(['transaction_id', 'transaction_date'], kind='mergesort')
header_cols = ['transaction_id','customer_id','payment_method','location','transaction_date','discount_applied']
stg_txn_header = (
    stg_sorted
    .drop_duplicates(subset=['transaction_id'], keep='last')[header_cols]
    .copy()
)

# Enrich header with calendar parts for convenience
stg_txn_header['transaction_date'] = pd.to_datetime(stg_txn_header['transaction_date'], errors='coerce')
stg_txn_header['year']  = stg_txn_header['transaction_date'].dt.year.astype('Int64')
stg_txn_header['month'] = stg_txn_header['transaction_date'].dt.month.astype('Int64')
stg_txn_header['day']   = stg_txn_header['transaction_date'].dt.day.astype('Int64')
# SQLite weekday: 0=Sunday in strftime('%w'); we’ll compute in SQL later if needed


con = sqlite3.connect(DB_PATH)
cur = con.cursor()
cur.execute("PRAGMA foreign_keys = ON;")

ddl = """
-- Lookups
CREATE TABLE IF NOT EXISTS customers (
  customer_id        INTEGER PRIMARY KEY,
  customer_code      TEXT NOT NULL UNIQUE
);

CREATE TABLE IF NOT EXISTS categories (
  category_id        INTEGER PRIMARY KEY,
  category_name      TEXT NOT NULL UNIQUE
);

CREATE TABLE IF NOT EXISTS items (
  item_id            INTEGER PRIMARY KEY,
  item_code          TEXT NOT NULL UNIQUE,
  category_id        INTEGER NOT NULL,
  FOREIGN KEY (category_id) REFERENCES categories(category_id)
);

CREATE TABLE IF NOT EXISTS payment_methods (
  payment_method_id  INTEGER PRIMARY KEY,
  method_name        TEXT NOT NULL UNIQUE
);

CREATE TABLE IF NOT EXISTS channels (
  channel_id         INTEGER PRIMARY KEY,
  channel_name       TEXT NOT NULL UNIQUE
);

-- Transactions (header)
CREATE TABLE IF NOT EXISTS transactions (
  txn_id             INTEGER PRIMARY KEY,
  txn_code           TEXT NOT NULL UNIQUE,
  customer_id        INTEGER NOT NULL,
  payment_method_id  INTEGER NOT NULL,
  channel_id         INTEGER NOT NULL,
  txn_ts             TEXT NOT NULL,               -- ISO8601 date or datetime
  discount_applied   INTEGER NOT NULL DEFAULT 0,
  year               INTEGER,
  month              INTEGER,
  day                INTEGER,
  dow                INTEGER,                     -- 0=Sunday via strftime('%w')

  FOREIGN KEY (customer_id)       REFERENCES customers(customer_id),
  FOREIGN KEY (payment_method_id) REFERENCES payment_methods(payment_method_id),
  FOREIGN KEY (channel_id)        REFERENCES channels(channel_id)
);

-- Transaction items (detail)
CREATE TABLE IF NOT EXISTS transaction_items (
  txn_item_id        INTEGER PRIMARY KEY,
  txn_id             INTEGER NOT NULL,
  item_id            INTEGER NOT NULL,
  price_per_unit_c   INTEGER NOT NULL CHECK (price_per_unit_c >= 0),
  quantity           REAL    NOT NULL CHECK (quantity > 0),
  discount_amount_c  INTEGER NOT NULL DEFAULT 0 CHECK (discount_amount_c >= 0),
  discount_pct       REAL    NOT NULL DEFAULT 0 CHECK (discount_pct >= 0 AND discount_pct <= 100),
  line_total_c       INTEGER NOT NULL CHECK (line_total_c >= 0),
  FOREIGN KEY (txn_id)  REFERENCES transactions(txn_id) ON DELETE CASCADE,
  FOREIGN KEY (item_id) REFERENCES items(item_id)
);

-- Integrity triggers (line_total matches computed)
CREATE TRIGGER IF NOT EXISTS trg_ti_check_before_insert
BEFORE INSERT ON transaction_items
FOR EACH ROW
BEGIN
  SELECT
    CASE
      WHEN NEW.line_total_c !=
           CAST(ROUND(NEW.price_per_unit_c * NEW.quantity) AS INTEGER)
           - NEW.discount_amount_c
           - CAST(ROUND((NEW.discount_pct/100.0) * NEW.price_per_unit_c * NEW.quantity) AS INTEGER)
      THEN RAISE(ABORT, 'line_total_c does not match computed total')
    END;
END;

CREATE TRIGGER IF NOT EXISTS trg_ti_check_before_update
BEFORE UPDATE ON transaction_items
FOR EACH ROW
BEGIN
  SELECT
    CASE
      WHEN NEW.line_total_c !=
           CAST(ROUND(NEW.price_per_unit_c * NEW.quantity) AS INTEGER)
           - NEW.discount_amount_c
           - CAST(ROUND((NEW.discount_pct/100.0) * NEW.price_per_unit_c * NEW.quantity) AS INTEGER)
      THEN RAISE(ABORT, 'line_total_c does not match computed total')
    END;
END;

-- Indexes
CREATE INDEX IF NOT EXISTS idx_txn_ts                 ON transactions(txn_ts);
CREATE INDEX IF NOT EXISTS idx_txn_customer           ON transactions(customer_id);
CREATE INDEX IF NOT EXISTS idx_txn_payment_channel    ON transactions(payment_method_id, channel_id);
CREATE INDEX IF NOT EXISTS idx_ti_txn                 ON transaction_items(txn_id);
CREATE INDEX IF NOT EXISTS idx_ti_item                ON transaction_items(item_id);
CREATE INDEX IF NOT EXISTS idx_items_category         ON items(category_id);

-- Views (human-friendly dollars)
CREATE VIEW IF NOT EXISTS v_transaction_items AS
SELECT
  ti.txn_item_id,
  t.txn_code,
  i.item_code,
  c.category_name,
  (ti.price_per_unit_c / 100.0) AS price_per_unit,
  ti.quantity,
  (ti.discount_amount_c / 100.0) AS discount_amount,
  ti.discount_pct,
  (ti.line_total_c / 100.0)     AS line_total,
  t.txn_ts,
  pm.method_name   AS payment_method,
  ch.channel_name  AS channel,
  cust.customer_code AS customer
FROM transaction_items ti
JOIN transactions t     ON t.txn_id = ti.txn_id
JOIN items i            ON i.item_id = ti.item_id
JOIN categories c       ON c.category_id = i.category_id
JOIN payment_methods pm ON pm.payment_method_id = t.payment_method_id
JOIN channels ch        ON ch.channel_id = t.channel_id
JOIN customers cust     ON cust.customer_id = t.customer_id;

CREATE VIEW IF NOT EXISTS v_transactions AS
SELECT
  t.txn_id,
  t.txn_code,
  cust.customer_code AS customer,
  t.txn_ts,
  t.year, t.month, t.day, t.dow,
  pm.method_name     AS payment_method,
  ch.channel_name    AS channel,
  SUM(ti.line_total_c)/100.0 AS txn_total
FROM transactions t
JOIN customers cust     ON cust.customer_id = t.customer_id
JOIN payment_methods pm ON pm.payment_method_id = t.payment_method_id
JOIN channels ch        ON ch.channel_id = t.channel_id
LEFT JOIN transaction_items ti ON ti.txn_id = t.txn_id
GROUP BY t.txn_id;
"""
cur.executescript(ddl)
con.commit()
print("Schema created in", DB_PATH)
stg_sales_raw_cols = [
    'transaction_id','customer_id','category','item',
    'price_per_unit','quantity','total_spent',
    'payment_method','location','transaction_date',
    'discount_applied','discount_amount','discount_percentage'
]
stg_to_sql = stg[stg_sales_raw_cols].copy()

# SQLite likes plain strings for dates; keep ISO format
stg_to_sql['transaction_date'] = pd.to_datetime(stg_to_sql['transaction_date']).dt.strftime('%Y-%m-%d')

# Write/replace staging tables
stg_to_sql.to_sql('stg_sales_raw', con, if_exists='replace', index=False)
stg_txn_header.to_sql('stg_txn_header', con, if_exists='replace', index=False)
print("Staging tables loaded: stg_sales_raw, stg_txn_header")
#Populate dimension tables (INSERT OR IGNORE)
cur = con.cursor()

cur.executescript("""
INSERT OR IGNORE INTO customers (customer_code)
SELECT DISTINCT customer_id FROM stg_txn_header WHERE customer_id IS NOT NULL;

INSERT OR IGNORE INTO categories (category_name)
SELECT DISTINCT category FROM stg_sales_raw WHERE category IS NOT NULL;

INSERT OR IGNORE INTO payment_methods (method_name)
SELECT DISTINCT payment_method FROM stg_sales_raw WHERE payment_method IS NOT NULL;

INSERT OR IGNORE INTO channels (channel_name)
SELECT DISTINCT location FROM stg_sales_raw WHERE location IS NOT NULL;
""")

# Items depend on category_id; do with a join
cur.executescript("""
INSERT OR IGNORE INTO items (item_code, category_id)
SELECT DISTINCT s.item,
       c.category_id
FROM stg_sales_raw s
JOIN categories c ON c.category_name = s.category
WHERE s.item IS NOT NULL;
""")

con.commit()
print("Dimensions populated.")
# Insert transaction headers
cur.executescript("""
INSERT OR IGNORE INTO transactions (
  txn_code, customer_id, payment_method_id, channel_id, txn_ts, discount_applied, year, month, day, dow
)
SELECT
  h.transaction_id,
  (SELECT customer_id FROM customers WHERE customer_code = h.customer_id),
  (SELECT payment_method_id FROM payment_methods WHERE method_name = h.payment_method),
  (SELECT channel_id FROM channels WHERE channel_name = h.location),
  h.transaction_date,
  h.discount_applied,
  h.year, h.month, h.day,
  CAST(strftime('%w', h.transaction_date) AS INTEGER)
FROM stg_txn_header h
WHERE h.transaction_id IS NOT NULL;
""")

con.commit()
print("Transaction headers inserted.")
# Insert transaction Items
# Build a view-like temp table with cents conversions via pandas for speed & clarity
ti = stg[['transaction_id','item','price_per_unit','quantity','discount_amount','discount_percentage','total_spent']].copy()

# Filter valid lines: need transaction_id and item, quantity > 0, price >= 0
ti = ti[ti['transaction_id'].notna() & ti['item'].notna()].copy()
ti['quantity'] = pd.to_numeric(ti['quantity'], errors='coerce')
ti['price_per_unit'] = pd.to_numeric(ti['price_per_unit'], errors='coerce')
ti['discount_amount'] = pd.to_numeric(ti['discount_amount'], errors='coerce').fillna(0.0)
ti['discount_percentage'] = pd.to_numeric(ti['discount_percentage'], errors='coerce').fillna(0.0)
ti['total_spent'] = pd.to_numeric(ti['total_spent'], errors='coerce').fillna(0.0)

ti = ti[(ti['quantity'] > 0) & (ti['price_per_unit'] >= 0)]

# Convert to cents
ti['price_per_unit_c'] = ti['price_per_unit'].apply(to_cents)
ti['discount_amount_c'] = ti['discount_amount'].apply(to_cents)
ti['line_total_c']      = ti['total_spent'].apply(to_cents)

# Write a temp table for bulk insert via SQL join
ti.to_sql('stg_ti', con, if_exists='replace', index=False)

# Insert with FK mapping + let the trigger validate math
cur = con.cursor()
cur.executescript("""
INSERT INTO transaction_items (
  txn_id, item_id, price_per_unit_c, quantity, discount_amount_c, discount_pct, line_total_c
)
SELECT
  t.txn_id,
  i.item_id,
  st.price_per_unit_c,
  st.quantity,
  st.discount_amount_c,
  COALESCE(st.discount_percentage, 0.0),
  st.line_total_c
FROM stg_ti st
JOIN transactions t ON t.txn_code = st.transaction_id
JOIN items i        ON i.item_code = st.item;
""")
con.commit()

# Clean up the temp table
cur.execute("DROP TABLE IF EXISTS stg_ti;")
con.commit()
print("Transaction items inserted.")
print("Validation and Display below.")
q = {}

q['counts'] = """
SELECT 
 (SELECT COUNT(*) FROM customers)        AS customers,
 (SELECT COUNT(*) FROM categories)       AS categories,
 (SELECT COUNT(*) FROM items)            AS items,
 (SELECT COUNT(*) FROM payment_methods)  AS payment_methods,
 (SELECT COUNT(*) FROM channels)         AS channels,
 (SELECT COUNT(*) FROM transactions)     AS transactions,
 (SELECT COUNT(*) FROM transaction_items) AS transaction_items;
"""

q['sample_txn'] = """
SELECT * FROM v_transactions
ORDER BY txn_ts DESC
LIMIT 10;
"""

q['top_items'] = """
SELECT category_name, item_code, SUM(line_total) AS revenue
FROM v_transaction_items
GROUP BY category_name, item_code
ORDER BY revenue DESC
LIMIT 10;
"""

for name, sql in q.items():
    print(f"\n--- {name} ---")
    try:
        display(pd.read_sql(sql, con))
    except Exception as e:
        print("Error:", e)


Schema created in sales.db
Staging tables loaded: stg_sales_raw, stg_txn_header
Dimensions populated.
Transaction headers inserted.
Transaction items inserted.
Validation and Display below.

--- counts ---


customers  categories  items  payment_methods  channels  transactions  \
0         25           8    200                3         2         12575   

   transaction_items  
0              11362


--- sample_txn ---


txn_id     txn_code customer               txn_ts  year  month  day  dow  \
0     722  TXN_1505827  CUST_13  2025-01-18 00:00:00  2025      1   18    6   
1     949  TXN_1661883  CUST_04  2025-01-18 00:00:00  2025      1   18    6   
2    1099  TXN_1777313  CUST_07  2025-01-18 00:00:00  2025      1   18    6   
3    6012  TXN_5321918  CUST_24  2025-01-18 00:00:00  2025      1   18    6   
4    6577  TXN_5709336  CUST_10  2025-01-18 00:00:00  2025      1   18    6   
5    6704  TXN_5804265  CUST_11  2025-01-18 00:00:00  2025      1   18    6   
6    6841  TXN_5907338  CUST_25  2025-01-18 00:00:00  2025      1   18    6   
7    7500  TXN_6383632  CUST_12  2025-01-18 00:00:00  2025      1   18    6   
8    8867  TXN_7369318  CUST_01  2025-01-18 00:00:00  2025      1   18    6   
9      19  TXN_1011669  CUST_14  2025-01-17 00:00:00  2025      1   17    5   

   payment_method   channel  txn_total  
0  Digital Wallet    Online        NaN  
1            Cash  In-Store      261.0  
2     Credit Card    Online        NaN  
3     Credit Card  In-Store       62.5  
4     Credit Card  In-Store       19.0  
5  Digital Wallet    Online       67.0  
6            Cash  In-Store      342.0  
7            Cash  In-Store       70.0  
8     Credit Card    Online      100.0  
9     Credit Card    Online      184.0


--- top_items ---


category_name     item_code  revenue
0                      Furniture   Item_25_FUR  25256.0
1  Electric Household Essentials   Item_25_EHE  23083.0
2                       Butchers   Item_25_BUT  21894.0
3                      Furniture   Item_24_FUR  21172.0
4                           Food  Item_25_FOOD  20541.0
5                       Butchers   Item_22_BUT  19710.0
6                       Butchers   Item_23_BUT  19114.0
7                       Butchers   Item_20_BUT  18860.5
8                  Milk Products  Item_19_MILK  18848.0
9  Electric Household Essentials   Item_23_EHE  18468.0